src/Advection/weno.jl

```diff
@@ -0,0 +1,125 @@
+#####
+##### Weighted Essentially Non-Oscillatory (WENO) schemes
+#####
+
+struct WENO{K, FT, C, B, Γ} <: AbstractAdvectionScheme
+    C :: C
+    B :: B
+    γ :: Γ
+    ε :: FT
+    η :: FT
+end
+
+function WENO(FT, n)
+    (n < 3 || iseven(n)) &&
+        error("WENO schemes are only defined for order n = 3, 5, 7, 9, 11, ...")
+
+    n >= 7 &&
+        @info "Computing WENO-$n smoothness indicator coefficients. This could take a while for n >= 7..."
+
+    k = Int((n + 1) / 2)
+    C = eno_coefficients_matrix(FT, k)
+    B = β_coefficients(FT, k)
+    γ = optimal_weno_weights(FT, k)
+
+    ε = 1e-6  # To avoid dividing by zero
+    η = 2     # WENO exponent
+
+    return WENO{k, FT, typeof(C), typeof(B), typeof(γ)}(C, B, γ, ε, η)
+end
+
+WENO(n) = WENO(Float64, n)
+
+#####
+##### ENO interpolants
+#####
+
+@inline px(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.C[ℓ+1, r+1] * ϕ[i-r-1+ℓ, j, k] for ℓ in 0:K-1)
+@inline py(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.C[ℓ+1, r+1] * ϕ[i, j-r-1+ℓ, k] for ℓ in 0:K-1)
+@inline pz(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.C[ℓ+1, r+1] * ϕ[i, j, k-r-1+ℓ] for ℓ in 0:K-1)
+
+#####
+##### Jiang & Shu (1996) WENO smoothness indicators
+#####
+
+@inline βx(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.B[m+1, n+1, r+1] * ϕ[i-r-1+m, j, k] * ϕ[i-r-1+n, j, k] for m in 0:K-1 for n in 0:m)
+@inline βy(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.B[m+1, n+1, r+1] * ϕ[i, j-r-1+m, k] * ϕ[i, j-r-1+n, k] for m in 0:K-1 for n in 0:m)
+@inline βz(i, j, k, weno::WENO{K}, r, ϕ) where K = @inbounds sum(weno.B[m+1, n+1, r+1] * ϕ[i, j, k-r-1+m] * ϕ[i, j, k-r-1+n] for m in 0:K-1 for n in 0:m)
+
+#####
+##### WENO raw weights
+#####
+
+@inline αx(i, j, k, weno, r, ϕ) = @inbounds weno.γ[r+1] / (βx(i, j, k, weno, r, ϕ) + weno.ε)^weno.η
+@inline αy(i, j, k, weno, r, ϕ) = @inbounds weno.γ[r+1] / (βy(i, j, k, weno, r, ϕ) + weno.ε)^weno.η
+@inline αz(i, j, k, weno, r, ϕ) = @inbounds weno.γ[r+1] / (βz(i, j, k, weno, r, ϕ) + weno.ε)^weno.η
+
+#####
+##### WENO normalized weights
+#####
+
+@inline function weno_weights_x(i, j, k, weno::WENO{K}, ϕ) where K
+    α = [αx(i, j, k, weno, r, ϕ) for r in 0:K-1]
```

Comment:
I'm not sure the array generation gets inlined away. I think a tuple generator may be better (but I'm not sure).

Reply:
True. Will change it.